packages/yarnpkg-core/sources/Cache.ts

```diff
@@ -331,14 +337,21 @@ export class Cache {
         if (!isCacheHit) {
           return loadPackage();
         } else {
-          let checksum: string | null = null;
+          let checksum: string | null = null
           const cachePath = tentativeCachePath!;
-
-          if (!shouldMock)
-            checksum = this.check
-              ? await validateFileAgainstRemote(cachePath)
-              : await validateFile(cachePath);
-
+          if (!shouldMock) {
+            if (this.check) {
+              checksum = await validateFileAgainstRemote(cachePath);
+            } else {
+              const mbChecksum = await validateFile(cachePath);
```

Comment:
mb instead of changing `validateFile` body, just check it here:
```js
if (checksum !== expectedChecksum && behavuour === 'reset') {  return loadPackage() } 
```